text/0002-interfaces.md

```diff
@@ -0,0 +1,300 @@
+      - a warning may be emitted if the `.shape` of the member and of the attribute are different
+    - if the member is an interface, the attribute value must be compatible with the signature of the member
+  * `signature.__call__()` freezes this signature and creates an interface from it. This method is expected to be routinely overridden in `Signature` subclasses to perform actions specific to a particular signature. The interface is created as follows:
+    1. A fresh Python object `obj` is created using `object()`
+    2. `obj.signature` is set to `signature`
+    3. For every member of the signature, an attribute with the member's name is set on `obj`, where the value is:
+      * If the member is a port, `Signal(member.shape, reset=member.reset)`.
+      * If the member is a signature, `member.signature.__call__()`.
+
+
+### Interface connection
+
+Interfaces may be connected to each other using the `amaranth.lib.module.connect(m, first, *rest)` free function.
+
+This function connects interfaces that satisfy the following requirements:
+* For every port member with a given `path` and `first_shape` in `first`, there is a port member with the same `path` in each of `rest` with shape `rest_shape`, where `Shape.cast(first_shape) == Shape.cast(rest_shape)`, and there are no other members in any of `rest`. In other words, the width and signedness of all of port members must be equal, but the shape-castable objects specifying the width and signedness do not have to be.
+  * It is expected that any mismatch in signatures will be resolved (through wrappers, mutating signature members, or otherwise) before interfaces are being connected.
+* Either:
+  1. There is exactly one interface in `rest`, and for every pair of members `first_member` and `rest_member`, `first_member.flow == ~rest_member.flow`.
+  2. There is any amount of interfaces in `rest`, and for every pair of members `first_member` and `rest_member`, `first_member.flow == Out` and `rest_member.flow == In`.
+
+In both of the cases (1) and (2), the function, for each pair of an input port and an output of port with the same path, the function connects these as follows:
+```python
+m.d.comb += output_port.eq(input_port)
+```
+
+
+## Drawbacks
+
+- The author considers `amaranth.lib.module` to be the best choice for this module, but it is ambiguous with `amaranth.hdl.dsl.Module`. In the author's view, if this ambiguity is a problem, then it is best to rename the latter (to `Builder`, since it is building up a netlist).
+
+
+## Alternatives and rationale
+
+- Do nothing. `Record` will continue to be used alongside the continued proliferation of ad-hoc implementations of similar functionality, and continue to impair the use of Amaranth components together.
+
+- Replace the `interface.lib.module.connect` free function with a function `Interface.connect` or a function `amaranth.hdl.dsl.Module.connect`.
```

Comment:
```suggestion
- Replace the `amaranth.lib.module.connect` free function with a function `Interface.connect` or a function `amaranth.hdl.dsl.Module.connect`.
```

Also, this RFC no longer proposes an `Interface` class, and `Signature.connect` wouldn't work.